### DISCUSSION TITLE
Question on air freshener detection and counting on conveyor belt using Yolov8s

### DISCUSSION THREAD

**Discussion by acse-yw3523** [ultralytics/ultralytics#14364] [Q&A] [answered]
Hi,

I'm working on a student project to detect and count air fresheners on a factory conveyor belt using the YOLOv8s model for object detection. There are 14 different scents of air fresheners. The company provided me with about 10 images for each category, so I have a total of 126 images for the YOLOv8s train/val/test set.

I think the number of images might be insufficient, even with data augmentation, so I'm considering transfer learning. I plan to train the pre-trained YOLOv8s model first with 3,000 images of different bottles because I believe the shape and texture of bottles might be similar to the air freshener containers. Then, I intend to save this model and use it for a second round of transfer learning to train on the air freshener dataset.

Does this approach seem reasonable? Will performing transfer learning twice help me achieve better model results?

Also, could you advise on which step should I use the model.tune function, should I do it after I finished trainning the air fresheners dataset to prevent overfitting?

Thank you!

**Comment by acse-yw3523** [comment]
![results](https://github.com/ultralytics/ultralytics/assets/142487377/976311a8-5ab1-4c6a-93c1-acc7d5c752dd)
BTW this is the final result.

**Reply by glenn-jocher** [reply]
@acse-yw3523 hi there,

Thank you for sharing your project details and the final result image! Your approach to use transfer learning twice is indeed a reasonable strategy, especially given the limited number of images in your air freshener dataset. Here are a few points to consider:

1. **Initial Transfer Learning with Bottles**: Training the YOLOv8s model on a larger dataset of different bottles can help the model learn general features related to shape and texture, which can be beneficial when fine-tuning on your specific air freshener dataset.

2. **Second Round of Transfer Learning**: Fine-tuning the model on your air freshener dataset after the initial training can help the model adapt to the specific characteristics of the air fresheners. This two-step transfer learning process can indeed improve the model's performance.

3. **Data Augmentation**: Given the limited number of images, data augmentation will be crucial. Techniques such as rotation, scaling, and color jittering can help create a more robust model.

4. **Model Tuning**: The `model.tune` function can be used to fine-tune hyperparameters to prevent overfitting. You can apply this after training on the air freshener dataset to optimize the model's performance further.

Here's a brief code snippet to illustrate the process:

```python
from ultralytics import YOLO

# Step 1: Train on the bottle dataset
model = YOLO('yolov8s.pt')
model.train(data='bottle_dataset.yaml', epochs=50, imgsz=640)

# Save the intermediate model
model.save('bottle_trained_model.pt')

# Step 2: Fine-tune on the air freshener dataset
model = YOLO('bottle_trained_model.pt')
model.train(data='air_freshener_dataset.yaml', epochs=50, imgsz=640)

# Optional: Tune the model to prevent overfitting
model.tune(data='air_freshener_dataset.yaml')
```

Feel free to adjust the number of epochs and other hyperparameters based on your observations during training. Best of luck with your project, and feel free to share any further questions or updates! 🚀

**Reply by acse-yw3523** [reply]
Hi @glenn-jocher , thank you so much! I'm trying both (train directly on the yolov8s & do transfer learning twice), but I don't know why it seems like train on the yolov8s use air fresheners dataset directly provides a better result on prediction, even when training on the bottle dataset, I used a variety of bottles, such as plastic, glass, and transparent ones, to give the model generalization ability. 

The trainning results of 'train directly on the yolov8'⬇️
![results_1TL](https://github.com/user-attachments/assets/30d08ae5-b7b3-492c-a116-cd40d2526205)

The trainning results of 'do transfer learning twice'⬇️
![results_2TL](https://github.com/user-attachments/assets/251a587b-3c89-400c-abbb-7a7408e33105)


BTW, my understanding is that model.tune is also training the model simultaneously, so the output in the ‘runs’ folder at the end will be the best model and the best hyperparameters? Is my understanding correct? 
Or do I need to put the obtained best hyperparameters .yaml back into the model for retraining (meaning, first step: train on the bottle dataset, second step: train on the air freshener dataset, third step: tune the model output from the air freshener dataset, fourth step: use the fine-tuned hyperparameters to retrain the model on the air freshener dataset again)?

Thank you again for your help!

**Reply by glenn-jocher** [reply]
Hi @acse-yw3523,

Thank you for sharing your results and observations! It's great to see your proactive approach in experimenting with different training strategies.

Regarding your question about `model.tune`, your understanding is partially correct. The `model.tune` function indeed helps in fine-tuning hyperparameters, but it does not retrain the model from scratch. Instead, it optimizes the hyperparameters to improve the model's performance based on the validation set.

Here's a concise breakdown of the steps you can follow:

1. **Initial Training on Bottles**: Train the model on the bottle dataset.
2. **Fine-Tuning on Air Fresheners**: Train the model on the air freshener dataset.
3. **Hyperparameter Tuning**: Use `model.tune` to optimize hyperparameters based on the air freshener dataset.
4. **Final Training (if needed)**: Optionally, retrain the model using the best hyperparameters obtained from the tuning step.

```python
from ultralytics import YOLO

# Step 1: Train on the bottle dataset
model = YOLO('yolov8s.pt')
model.train(data='bottle_dataset.yaml', epochs=50, imgsz=640)

# Step 2: Fine-tune on the air freshener dataset
model.train(data='air_freshener_dataset.yaml', epochs=50, imgsz=640)

# Step 3: Tune hyperparameters
model.tune(data='air_freshener_dataset.yaml')

# Optional Step 4: Retrain with tuned hyperparameters
model.train(data='air_freshener_dataset.yaml', epochs=50, imgsz=640, hyp='best_hyp.yaml')
```

This approach ensures that you leverage the optimized hyperparameters for the best possible performance. 

If you encounter any issues, please ensure you are using the latest versions of the packages and provide a reproducible example if needed. You can find more details on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Best of luck with your project! 🚀

**Reply by acse-yw3523** [reply]
Hi @glenn-jocher, Thank you so much for your explanation! 

May I also ask if it is reasonable to follow these steps?

**Step 1:** Train the bottle dataset using YOLOv8s and save the output model (yolov8s_bottle.pt)

**Step 2:** Use yolov8s_bottle.pt to perform hyperparameter tuning on the air freshener dataset:
`model = YOLO("yolov8s_bottle.pt")
model.tune(data="air_fresheners.yaml", epochs=30, iterations=50, optimizer="AdamW", plots=False, save=False, val=False)
`

**Step 3:** Use the obtained best_hyperparameters.yaml to train from scratch:
`model = YOLO("yolov8s_bottle.pt")
model.train(data='air_fresheners.yaml', cfg='best_hyperparameters.yaml', epochs=100)
`

**Reply by glenn-jocher** [reply, marked as answer]
Hi @acse-yw3523,

Your proposed steps look solid and well thought out! Here's a concise breakdown:

1. **Train on Bottle Dataset**: Train the YOLOv8s model on the bottle dataset and save the output model (`yolov8s_bottle.pt`).
2. **Hyperparameter Tuning**: Use `yolov8s_bottle.pt` to perform hyperparameter tuning on the air freshener dataset.
3. **Final Training**: Use the best hyperparameters obtained from tuning to train the model on the air freshener dataset.

Your code snippets are mostly correct, but you might want to ensure that the `model.tune` function saves the best hyperparameters. Here's a refined version:

```python
from ultralytics import YOLO

# Step 1: Train on the bottle dataset
model = YOLO('yolov8s.pt')
model.train(data='bottle_dataset.yaml', epochs=50, imgsz=640)
model.save('yolov8s_bottle.pt')

# Step 2: Hyperparameter tuning on the air freshener dataset
model = YOLO('yolov8s_bottle.pt')
model.tune(data='air_fresheners.yaml', epochs=30, iterations=50, optimizer='AdamW', plots=False, save=True, val=True)

# Step 3: Final training with best hyperparameters
model.train(data='air_fresheners.yaml', hyp='best_hyperparameters.yaml', epochs=100)
```

This approach should help you achieve optimal performance. If you encounter any issues, please ensure you are using the latest versions of the packages. For any reproducible issues, you can refer to [this guide](https://docs.ultralytics.com/help/minimum_reproducible_example).

Best of luck with your project! 🚀

**Reply by acse-yw3523** [reply]
Thank you!

**Reply by glenn-jocher** [reply]
You're welcome! If you have any more questions or run into any issues, feel free to ask. We're here to help!
